internal/registry/registry.go

```diff
@@ -0,0 +1,219 @@
+		}
+		return getMultiPlatformManifests(ctx, client, r.baseURL, &index)
+	} else if _, ok = mediaTypeImageManifestV1[mediaType]; ok {
+		manifest := imageManifestV1{}
+		if err = json.Unmarshal(b, &manifest); err != nil {
+			return nil, fmt.Errorf("error unmarshalling image manifest from %s: %w", url, err)
+		}
+		manifest.URL = url
+		return []*imageManifestV1{&manifest}, nil
+	} else {
+		return nil, fmt.Errorf("unknown mediaType %s from %s: %w", mediaType, url, err)
+	}
+}
+
+func getMultiPlatformManifests(ctx context.Context, client httpclient.HTTPClient, baseURL string, index *imageIndexV1) ([]*imageManifestV1, error) {
+	var manifests = make([]*imageManifestV1, len(index.Manifests))
+	for i, ref := range index.Manifests {
+		url := fmt.Sprintf("%s/manifests/%s", baseURL, ref.Digest)
+		manifest := imageManifestV1{}
+		if err := client.GetJSON(ctx, url, ref.MediaType, &manifest); err != nil {
+			return nil, fmt.Errorf("error getting image ref for platform %v: %w", ref.Platform, err)
+		}
+		manifest.URL = url
+		manifests[i] = &manifest
+	}
+	return manifests, nil
+}
+
+func (r *registry) getImageConfigs(ctx context.Context, client httpclient.HTTPClient, images []*imageManifestV1) ([]*imageConfigV1, error) {
+	var configs = make([]*imageConfigV1, len(images))
+	for i, image := range images {
+		if !isMediaTypeImageConfigV1(image.Config.MediaType) {
+			return nil, fmt.Errorf("invalid config media type in image %v", image)
+		}
+		url := fmt.Sprintf("%s/blobs/%s", r.baseURL, image.Config.Digest)
+		config := imageConfigV1{}
+		err := client.GetJSON(ctx, url, image.Config.MediaType, &config)
+		if err != nil {
+			return nil, fmt.Errorf("error getting image config from %s: %w", url, err)
+		}
+		configs[i] = &config
+	}
+	return configs, nil
+}
+
+func (r *registry) ReadFilesystemLayer(ctx context.Context, layer *internal.FilesystemLayer, readFile internal.ReadFile) error {
+	client := r.httpClient(ctx)
+	body, _, err := client.Get(ctx, layer.URL, &http.Header{"Accept": []string{layer.MediaType}})
+	if err != nil {
+		return err
+	}
+	zSrc, err := gzip.NewReader(body)
+	if err != nil {
+		return err
+	}
+	defer zSrc.Close() //nolint
+
+	tr := tar.NewReader(zSrc)
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		// Skip directories, symbolic links, block devices, etc.
+		if header.Typeflag != tar.TypeReg {
+			continue
+		}
+
+		// We currently don't implement deleting files from the list
+		// https://github.com/opencontainers/image-spec/blob/859973e32ccae7b7fc76b40b762c9fff6e912f9e/layer.md#whiteouts
+		if strings.Contains(header.Name, ".wh.") {
+			continue
+		}
+		if err := readFile(header.Name, header.Size, header.Mode, header.ModTime, tr); err != nil {
+			return fmt.Errorf("error calling readFile on %s: %w", header.Name, err)
+		}
+	}
+	return nil
+}
+
+func (r *registry) httpClient(ctx context.Context) httpclient.HTTPClient {
```

Comment:
maybe we don't need this per comment https://github.com/tetratelabs/car/pull/3/files#r660327804